src/core/DataFile.cpp

```diff
@@ -1809,6 +1810,27 @@ void DataFile::upgrade_sampleAndHold()
 	}
 }
 
+//! Update MIDI CC indexes, so that they are counted from 0. Older releases of LMMS
+//! count the CCs from 1.
+void DataFile::upgrade_midiCCIndexing()
+{
+	std::vector<const char*> attributesToUpdate = {"inputcontroller", "outputcontroller"};
```

Comment:
Could use a `static constexpr std::array<const char*>` here, since the attributes don't change dynamically.

Reply:
If we use an initializer list we can even skip the tempate parameters: `static constexpr std::array`. Otherwise we'd have to specify the size as well: `static constexpr std::array<const char*, 2>`.

Reply:
The trouble with that is that string literals actually have type `const char[N]`, so different length strings have different types. I would lean towards specifying the dimension for now, and in the future we can have some user-defined literals that coerce the strings into consistent types.

Reply:
Actually, I think it's fine here. I've had some trouble like that in the past, but apparently not in this situation.